### DISCUSSION TITLE
about ecole cumulative time error in primal task

### DISCUSSION THREAD

**Discussion by superboom** [ds4dm/ml4co-competition#70] [Q&A] [answered]
when deploying official evaluation pipeline, we found the cumulative time was calculated with 6 instead of 5 mins. Therefore, the total reward will be added with shifted term which varied from submitted strategies.
Is that OK ?

![image](https://user-images.githubusercontent.com/30993400/131991839-c05d98e6-c6d8-424f-8801-53f6422ba6f7.png)
![image](https://user-images.githubusercontent.com/30993400/131993626-aff57745-c42d-430d-aa2e-6314832bdf0f.png)

**Comment by gasse** [comment]
Hi @superboom ,

That is strange indeed. Could you tell us at which place in the code you placed those print statements (Start and End) ? Also, which task is this one, primal, dual or config ? It might be that the environment spends exactly 1 min in reading + preprocessing the instance, which is not accounted for in the cumulative reward. But such a round number seems odd.

Best,
Maxime

**Reply by gasse** [reply]
Oh nevermind, 5 min is the primal task. Ok I'll look into it.

**Comment by gasse** [comment]
On our side it seems that the call to `env.reset(...)` indeed takes about 1.5 minutes for instance `load_balancing_9954`, but depending on your the hardware I guess 1 minute is plausible. See the execution trace below (note that I deactivated the loop over `env.step(...)` here so the cumulative reward is 0):
```
Instance load_balancing_9954.mps.gz
  seed: 0
  initial primal bound: 881.0
  initial dual bound: 689.4080272397499
  objective offset: 0
Time before reset: 76302.184255917
Time after reset: 76402.160111682
  cumulated reward (to be maximized): 0

Instance load_balancing_9954.mps.gz
  seed: 1
  initial primal bound: 881.0
  initial dual bound: 689.4080272397499
  objective offset: 0
Time before reset: 76402.160740505
Time after reset: 76479.576991936
  cumulated reward (to be maximized): 0

Instance load_balancing_9954.mps.gz
  seed: 2
  initial primal bound: 881.0
  initial dual bound: 689.4080272397499
  objective offset: 0
Time before reset: 76479.577707476
Time after reset: 76562.748952746
  cumulated reward (to be maximized): 0
```

The time spend in `reset` is not included in the time window _T=5min_ over which the primal integral is computed. Please let us know if those 6 minutes you measured include the call to `env.reset(...)` or not.

Best,
Maxime

**Comment by superboom** [comment]
Hi, Maxime
I think I didn't make a clear statement for my experiment.
I set 6 min sleep for policy class in agent.py, then the cumulated reward was calculated with 6 instead of 5, eventhougth set 'time_limit=5 min' for environment .
Usually, the policy run longer time limit than environment, which I cannot figure out.  I think total time = set time of scip + something I don't know. Therefore, I think there might be a shift term for total reward calculation

regards

**Comment by gasse** [comment, marked as answer]
Oh I see. Yes there is a corner case at the moment, when the agent spends too much time doing computations and overshoots the time limit set within the environment. Indeed, in that case the environment will stop _after_ the time limit has been met, and will compute the performance measures over a longer time window.

This can be mitigated on the agent side, by collecting from the environment the remaining running time and making sure that heavy computations are not run if there is not enough time left. See for example my suggested implementation here: https://github.com/ds4dm/ml4co-competition/discussions/31#discussioncomment-1043818

Handling this corner case on the environment side is doable in theory, but is not something trivial. I'll look into it, see if maybe we can handle that corner case properly [here](https://github.com/ds4dm/ml4co-competition/blob/33439dd4f93337d77b95aff71c666c412334828b/common/rewards.py#L48), but I can't guarantee anything.

Finally, please keep in mind that we also run the evaluation jobs themselves with a time limit, and that if that time limit is hit then the evaluation will simply result in a failure. It's really hard to recover in that case, since the process itself is killed by the system. So, all in all, it is a good idea to try to monitor and control heavy computations on the agent side :)

Hope that helps,

Best,
Maxime

**Reply by superboom** [reply]
Hi Maxime:

So I can set system break if policy shot time limit in my agent side, is that right?  Could it be acceptable for you if containing this operation in my submission?

it seems that your code in policy part extended time limitation for SCIP.  But this operation would increase total reward, is that ok?
![image](https://user-images.githubusercontent.com/30993400/132299264-1ff746a8-e963-4ac0-835d-e7190bbde393.png)

regards

**Reply by gasse** [reply]
Hi @superboom ,

I am not sure what you mean here by system break. I you mean killing the whole process, then no such an operation would not be recoverable on our side.

In the code you highlight, we set a time limit to a duplicated SCIP instance, which is managed by the agent and is completely separated from the SCIP instance within the environment. Hence, the agent runs a SCIP as a sub-routine with a certain time budget, which is computed based on the remaining time available for the whole process (agent+environment execution). This operation does not increase the total reward (the integral), on the opposite, it makes sure that the agent does not overshoots the time window over which we compute the integral.

Hope that helps,

Best,
Maxime
